File: enc.js
// Copyright (C) 2024 Code Eleven (Mustafa E / @codeeleven0)
// Hashing secrets are not for decoding. It protects hash collision.
const isEven = (x) => {
    return (x % 2) == 0;
}
const len = (x) => {
    return x.length;
}
const sum = (x) => {
    s = 0;
    x.forEach(h => s+=h);
    return s;
}
const div = (x, y) => {
    return Math.round(x / y);
} 
const ord = (x) => {
    return x.charCodeAt(0);
}
const chr = (x) => {
    return String.fromCharCode(x);
}
const max = (x) => {
    maxx = Number.MIN_VALUE;
    for(let i = 0; i < len(x); i++){
        if (x[i] > maxx){
            maxx = x[i];
        }
    }
    return maxx;
}
const min = (x) => {
    minn = Number.MAX_VALUE;
    for(let i = 0; i < len(x); i++){
        if (minn > x[i]){
            minn = x[i];
        }
    }
    return minn;
}
const findall = (pat, str) => {
    a = new RegExp(pat, "g");
    return str.match(a);
}
const strby = (s, by) => {
    ns = [];
    for(let i = 0; i < by; i++){
        ns.push(s)
    }
    return ns.join("");
}
const rjust = (str, times, d) => {
    ln = len(str);
    while(ln < times){
        str += d;
        ln = len(str);
        if (ln == times){
            break;
        }
    }
    return str;
}
const reversearr = (arr) => {
    return arr.reverse();
}
const calcAvg = (x) => {
    return div(sum(x), len(x));
}
const splitstring = (x) => {
    even = ""
    odd = ""
    for(let i = 0; i < len(x); i++){
        if (isEven(i)){
            even += x[i];
        }else{
            odd += x[i];
        }
    }
    return [even, odd];
}
const applysecret = (x, secret) => {
    splitted = splitstring(secret);
    encoded = [splitted[1] + x[0], splitted[0] + x[1]];
    return encoded;
}
const calculateintegrity = (x) => {
    s = x[0] + x[1];
    arr = [];
    for(let i = 0; i < len(s); i++){
        arr.push(ord(s[i]));
    }
    if (arr){
        sum1 = max(arr);
        sum2 = min(arr);
        sum3 = calcAvg(arr);
        chrs = "";
        for(let j = 33; j < 126; j++){
            chrs += chr(j);
        }
        if (chrs.includes(chr(sum3))){
            return [sum1, sum2, sum3];
        }else{
            return [sum1, sum2, arr[0]];
        }
    }else{
        return [65, 66, 67];
    }
}
const makeintegrity = (x) => {
    s = "";
    for(let j = 0; j < len(x); j++){
        s += chr(x[j]);
    }
    return s;
}
const makeend = (x) => {
    return x[0] + x[1];
}
const hashit = (x, secret) => {
    splitted = splitstring(x);
    encoded = applysecret(splitted, secret);
    integrity = calculateintegrity(encoded);
    integritystr = makeintegrity(integrity);
    hashstr = makeend(encoded);
    lastsplit = splitstring(hashstr + integritystr);
    hashstr2 = makeend(lastsplit);
    nstr = hashstr2.replaceAll(" ","*");
    n = "";
    if (len(nstr) < 15){
        n = nstr;
    }else{
        a = findall(strby(".",div(len(nstr), 15)), nstr);
        for(let s = 0; s < len(a); s++){
            ac = [];
            for (let i = 0; i < len(a[s]); i++){
                ac.push(ord(a[s][i]));
            }
            n += chr(div(sum(ac), len(ac)));
        }
    }
    if (len(n) < 15){
        try{
            n = rjust(n, 15, makeintegrity(calculateintegrity(splitstring(secret)))[-1]);
        }catch(e){
            n = rjust(n, 8, integritystr[-1]);
            n = rjust(n, 13, integritystr[0]);
            n = rjust(n, 15, integritystr[1]);
        }
    }
    if (len(n) > 15){
        n = findall(strby(".",15), n)[0];
    }
    try{
        end = makeend(reversearr(splitstring(n))) + chr(secret.charCodeAt(secret.length-1));
    }catch(x){
        end = makeend(reversearr(splitstring(n))) + chr(nstr.charCodeAt(nstr.length-1));
    }
    return end;
}
const _passwhash = (x, secret) => {
    try{
        return hashit(x, secret);
    }catch(e0){
        try{
            console.error(e0);
            return hashit("", secret);
        }catch(e1){
            try{
                console.error(e1);
                return hashit(x, "");
            }catch(e2){
                console.error(e2);
                return strby("0", 16);
            }
        }
    }
}
const passwhash = (x, secret) => {
    hr = _passwhash(x, secret).replaceAll(" ","*").replaceAll("\n","").replaceAll("\r","").replaceAll("\0","");
    if(len(hr) != 16){
        hr = rjust(hr, 16, hr[-1]);
    }
    return hr;
}
module.exports = {passwhash: passwhash}

File: enc.py
# Copyright (C) 2024 Code Eleven (Mustafa E / @codeeleven0)
# Hashing secrets are not for decoding. It protects hash collision.
import re
chs = 32
chss = chs - 1
def isEven(x: int) -> bool:
    if (x % 2) == 0:
        return True
    else:
        return False
def calcavg(x: list) -> int:
    return sum(x) // len(x)
def splitstring(x: str, secret: None = None) -> tuple[str, str]:
    even: str = "".join([x[c] for c in range(len(x)) if isEven(c)])
    odd: str = "".join([x[c] for c in range(len(x)) if not isEven(c)])
    return (even, odd)
def applysecret(x: tuple[str, str], secret: str) -> tuple[str, str]:
    splitted: tuple[str, str] = splitstring(secret)
    encoded: tuple[str, str] = (splitted[1] + x[0], splitted[0] + x[1])
    return encoded
def calculateintegrity(x: tuple[str, str]) -> tuple[int, int, int]:
    arr: list = [ord(c) for c in "".join(x)]
    if arr:
        sum1: int = max(arr)
        sum2: int = min(arr)
        sum3: int = calcavg(arr)
        chrs: str = "".join([chr(i) for i in range(33, 126)])
        if chr(sum3) in chrs:
            return sum1, sum2, sum3
        else:
            return sum1, sum2, arr[0]
    else:
        return 65, 66, 67
def makeintegrity(x: tuple[int, int, int]) -> str:
    return "".join([chr(c) for c in x])
def makeend(x: tuple[str, str]) -> str:
    return x[0] + x[1]
def hashit(x: str, secret: str) -> str:
    splitted: tuple[str, str] = splitstring(x)
    encoded: tuple[str, str] = applysecret(splitted, secret)
    integrity: tuple[int, int, int] = calculateintegrity(encoded)
    integritystr: str = makeintegrity(integrity)
    hashstr: str = makeend(encoded)
    lastsplit: tuple[str, str] = splitstring(hashstr + integritystr)
    hashstr2: str = makeend(lastsplit)
    nstr: str = hashstr2.replace(" ", "*")
    n: str = ""
    if len(nstr) < chss:
        n = nstr
    else:
        for s in re.findall("."*(len(nstr)//chss), nstr):
            a: list = [ord(c) for c in s]
            n += chr(sum(a) // len(a))
    if len(n) < chss:
        try:
            n = n.ljust(chss,  makeintegrity(calculateintegrity(splitstring(secret)))[-1])
        except:
            n = n.ljust(chs//2, integritystr[-1])
            n = n.ljust(chs // 2 + chs // 3,  integritystr[0])
            n = n.ljust(chss, integritystr[1])
    if len(n) > chss:
        n = re.findall("."*chss, n)[0]
    try:
        end: str = makeend(splitstring(n)[-1::-1]) + secret[-1]
    except:
        end: str = makeend(splitstring(n)[-1::-1]) + nstr[-1]
    return end
def _passwhash(x: str, secret: str):
    try:
        return hashit(x, secret)
    except:
        try:
            return hashit("", secret)
        except:
            try:
                return hashit(x, "")
            except:
                return chs*"0"
def __passwhash(x: str, secret: str) -> str:
    hr: str = _passwhash(x, secret).replace(" ","*").replace("\n","").replace("\0","").replace("\r","")
    if len(hr) != chs:
        hr = hr.ljust(chs, hr[-1])
    return hr
def toHex(s: str):
    a = "".join([hex(ord(c) // 1).split("0x")[-1] for c in s])
    return a
def passwhash(data: str, secret: str):
    hashs = toHex(__passwhash(data, secret))
    return hashs